### ISSUE TITLE
Can I directly obtain the logits here?

### ISSUE THREAD

**Issue by SparkJiao** [vllm-project/vllm#185]
Hi, wonderful work!

I want to know if there is a easy way to obtain the logits, since sometimes I only need to calculate the perplexity/language modeling loss of specific sequence.

I saw the code here: [https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L211-L235](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L211-L235)

So I want to know if I directly use the logits generated by `lm_head`, can I be benefited from the paged attention framework? Thanks very much!

**Comment by zhuohan123** [comment]
For sequence generation, you can get the logits of the generated tokens by setting `log_probs` option in [SamplingParams](https://github.com/vllm-project/vllm/blob/67d96c29fba9b72cb4c4edbc26211c208a00ebdd/vllm/sampling_params.py#L40).

In the current vLLM, we haven't added support for getting the logits of the prompt parts yet. Theoretically, your specific use case can benefit from the no-padding batching of different-length sequences in vLLM, so vLLM can still be faster than plain huggingface transformers. However, since you only compute on prompts, there is no "KV Cache" and thus the memory management part of PagedAttention does not increase throughput.

**Comment by SparkJiao** [comment]
I see. Thansk for your clarification very much! I will try some cases first

**Comment by jpiabrantes** [comment]
@zhuohan123 

> For sequence generation, you can get the logits of the generated tokens by setting `log_probs` option in [SamplingParams](https://github.com/vllm-project/vllm/blob/67d96c29fba9b72cb4c4edbc26211c208a00ebdd/vllm/sampling_params.py#L40).

Isn't a `log_prob` different from a `logit`? `log_probs[i] = ln(e^logits[i]/sum(e^logits))`.

We can't recover the top-10 `logits` from the top-10 `log_probs` because we don't know what is the entire sum.

If this is the case can this issue be re-open?

**Comment by jpiabrantes** [comment]
> For sequence generation, you can get the logits of the generated tokens by setting `log_probs` option in [SamplingParams](https://github.com/vllm-project/vllm/blob/67d96c29fba9b72cb4c4edbc26211c208a00ebdd/vllm/sampling_params.py#L40).

`logits` are different from `logprobs`. `log_probs = log(exp(logits)/sum(exp(logits)))`

We can't recover the top 10 logits from the top 10 logprobs. Can this issue be re-open?

**Comment by junzhang-zj** [comment]
can i obtain the logits of the prompt, like the loss of evaluation？

**Comment by vmoens** [comment]
Is there any update on this? Has anyone figured out how to retrieve logits for a given input?

**Comment by metasyn** [comment]
@zhuohan123 can you see the various comments here, discussing how this was closed prematurely? i think we need to re-open this please 🙇

**Comment by universuen** [comment]
Hi everyone, I may find a tricky way to to this. vllm sampling params accept a customized `logits_processor`, which has a view of the logits. So we can use a `logits_processor` to attach the logits to a global variable and read it elsewhere. E.g.:

```python
import os

import torch
from vllm import LLM, SamplingParams


os.environ["VLLM_USE_V1"] = "0"  # vLLM V1 does not support logits processors.


class LogitsSpy:
    def __init__(self):
        self.processed_logits: list[torch.Tensor] = []
    def __call__(self, token_ids: list[int], logits: torch.Tensor):
        self.processed_logits.append(logits)
        return logits


if __name__ == "__main__":
    llm = LLM(model="Qwen/Qwen2.5-0.5B-Instruct")
    prompts = [
        "What is the capital of France?",
    ]
    logits_spy = LogitsSpy()
    sampling_params = SamplingParams(temperature=0.7, top_p=0.95, max_tokens=100, logits_processors=[logits_spy])
    outputs = llm.generate(prompts, sampling_params)
    print(f"Logits: {logits_spy.logits}")

```

**Comment by fscdc** [comment]
Hi @universuen ,
Have u ever check this when it comes to batch inference, i find it is a [(token_count+2), 15w] in single sample inference

**Comment by universuen** [comment]
> Hi [@universuen](https://github.com/universuen) , Have u ever check this when it comes to batch inference, i find it is a [(token_count+2), 15w] in single sample inference

No, I haven't. I suggest you not to inference in a batch manner, since this may cause inconsistent and decreased performance according to [this issuse](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535).

**Comment by fscdc** [comment]
Hi @universuen 
I try this way, but it may cause some problem for generation, such as only generate several token. i do not know the reason

**Comment by windygoo** [comment]
Any update?

**Comment by npip99** [comment]
@windygoo This is supported, https://github.com/vllm-project/vllm/blob/e952eee698fb8a497baaa864b93387fd180f27e7/docs/usage/v1_guide.md?plain=1#L162

**Comment by YunruiZhang** [comment]
> [@windygoo](https://github.com/windygoo) Getting the raw logits is supported,
> 
> [vllm/docs/usage/v1_guide.md](https://github.com/vllm-project/vllm/blob/e952eee698fb8a497baaa864b93387fd180f27e7/docs/usage/v1_guide.md?plain=1#L162)
> 
> Line 162 in [e952eee](/vllm-project/vllm/commit/e952eee698fb8a497baaa864b93387fd180f27e7)
> 
>  You can adjust this behavior by setting the `--logprobs-mode` flag.

Okay seems like the new implemation is buggy the    logprobs_mode = 'processed_logprobs', gives the same output as the  logprobs_mode = 'raw_logprobs' and the logits mode just does not work always gives -inf.
